hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlFetchMessageTask.java

```diff
@@ -59,7 +60,9 @@ protected ClientMessage encodeResponse(Object response) {
 
     @Override
     protected ClientMessage encodeException(Throwable throwable) {
-        nodeEngine.getSqlService().getInternalService().getClientStateRegistry().closeOnError(parameters.queryId);
+        // exception can be thrown before parameters are decoded
+        QueryId queryId = parameters != null ? parameters.queryId : decodeClientMessage(clientMessage).queryId;
+        nodeEngine.getSqlService().getInternalService().getClientStateRegistry().closeOnError(queryId);
```

Comment:
Same here, though the cursor likely already exists. I wonder what the exception was in this case, do you know?

Reply:
It was most probably one of the ones reported in #19319.
If we don't close the cursor here we leak memory, don't we?

Reply:
This most likely wasn't a normal exception. It happened before any code in this class executed. Therefore I think we should rely on the superclass handling:
```java
        // exception can be thrown before parameters are decoded
        if (parameters == null) {
            return super.encodeException(throwable);
        }
```